Feat/#45 마일스톤 CRUD 화면 구현, 서버와 api 연결 작업

sharpie1330 opened the pull request:
# 🚀 Pull Request

## 구현 내용
- 마일스톤 화면의 CRUD 기능을 구현했습니다.
- 서버와의 api 연동 작업을 했습니다. 그와 동시에 서버 코드에서 수정이 있어야 하는 부분도 함께 수정했습니다.

## 기타
- 마일스톤 서버 기능 중 수정되어야 할 부분들이 있습니다. 예외를 처리하는 부분 등이 그렇습니다.
- 해당 브랜치가 머지되고 나면 다른 브랜치에서 수정 작업을 진행합니다.

## 관련 이슈
close #45

sharpie1330 committed:
[feat/#45] feat: 마일스톤 스토어에 CRUD 로직 구현

sharpie1330 committed:
[feat/#45] feat: 마일스톤 필드 유효성 검증 함수 구현

sharpie1330 committed:
[feat/#45] refactor: 마일스톤 마감기한 필드 타입 Date로 변경

sharpie1330 committed:
[feat/#45] feat: 마일스톤 추가 폼 컴포넌트 구현

sharpie1330 committed:
[feat/#45] feat: 마일스톤 조회 화면 구현

sharpie1330 committed:
[feat/#45] feat: 마일스톤 수정 화면 구현

jellyyelly commented on frontend/src/components/milestone/MilestoneList.svelte:
머지되면 style 컴포넌트 통합이 급할것 같네요

jellyyelly commented on src/main/java/com/issuetracker/domain/milestone/Milestone.java:
```diff
@@ -26,7 +26,7 @@ public class Milestone extends BaseDateTime implements Persistable<String> {
     @Builder.Default
     private boolean isOpen = true;
-    private LocalDateTime dueDate;
+    private Date dueDate;
```
엇 Date 클래스는 java 8 이전 버전에서 제공하는 api 인데요, 레거시 때문에 사용하는게 아니라면 java.time 패키지의 `LocalDate`를 쓰는게 좋습니다!
그 이유에 대해선 gpt 형님의 답변이 있습니다.
```
Date 클래스
구 버전: java.util.Date 클래스는 Java 1.0부터 존재하는 오래된 API입니다.
설계 결함: Date 클래스는 불변이 아니며, 시간대를 제대로 처리하지 못하는 등의 여러 설계 결함이 있습니다.
메서드 폐기: 많은 메서드가 폐기(deprecated)되었고, 직관적이지 않은 방식으로 작동합니다.
LocalDate 클래스
Java 8 이후 도입: java.time 패키지는 Java 8에서 도입된 새로운 날짜와 시간 API의 일부입니다.
불변성: LocalDate는 불변 객체로, 값이 한 번 설정되면 변경되지 않습니다.
명확성: 날짜와 시간을 명확하고 직관적으로 다룰 수 있습니다. 예를 들어, LocalDate는 날짜만(연, 월, 일)을 다루며, LocalDateTime은 날짜와 시간까지 포함합니다.
풍부한 기능: 다양한 유틸리티 메서드를 제공하여 날짜 연산을 쉽게 할 수 있습니다.
타입 안정성: 더 명확한 타입을 제공하여 실수 가능성을 줄입니다.
```

jellyyelly commented on src/main/java/com/issuetracker/domain/milestone/MilestoneService.java:
꼼꼼하게 봐주셔서 고생하셨을게 보입니다.. ㅜ

jellyyelly commented on src/main/java/com/issuetracker/domain/milestone/request/MilestoneCreateRequest.java:
```diff
@@ -18,7 +18,7 @@ public class MilestoneCreateRequest {
     private String id;
 
-    private LocalDateTime dueDate;
+    private Date dueDate;
```
이 부분도 `LocalDate`로 변경하면 될 것 같습니다!

jellyyelly commented on src/main/java/com/issuetracker/domain/milestone/response/MilestoneResponse.java:
```diff
@@ -5,15 +5,15 @@
 
     private String id;
-    private LocalDateTime dueDate;
+    private Date dueDate;
```
앗 이 부분도요!

jellyyelly commented on src/test/java/com/issuetracker/domain/milestone/MilestoneControllerTest.java:
```diff
@@ -71,7 +73,7 @@ Stream<DynamicTest> create_validation() throws Exception {
         final String name = "n";
         final String description = "d";
-        final LocalDateTime dueDate = LocalDateTime.now();
+        final java.util.Date dueDate = Date.from(Instant.now());
```
이 부분도 `LocalDate`로 바꾸면 될 것 같습니다

jellyyelly commented on src/test/java/com/issuetracker/domain/milestone/MilestoneTest.java:
```diff
@@ -25,7 +25,7 @@ void create_success() {
         Milestone milestone = Milestone.builder()
                 .id("this is a test")
-                .dueDate(LocalDateTime.of(2999, 12, 31, 23, 59))
+                .dueDate(new Date(20240523))
```
이 부분도 있네요..!

jellyyelly reviewed (changes requested):
훌륭합니다 샤피!
한 가지 수정했으면 하는 부분은 리뷰에 달아뒀습니다!

sharpie1330 committed:
[feat/#45] fix: 마감기한 필드 타입 LocalDate로 변경

- Date 타입은 자바 1.0부터 존재하는 오래된 API로, 많은 메서드가 Deprecated되었으며 불변 객체가 아님. 따라서 java 8에서 새롭게 도입된 LocalDate를 사용.

sharpie1330 committed:
[feat/#45] fix: 마일스톤 조회 시 최신 생성 순으로 정렬되도록 구현 변경

sharpie1330 committed:
[feat/#45] feat: WebConfig에 모든 요청 메서드 허용 로직 추가

sharpie1330 committed:
[feat/#45] fix: 마감기한 값이 있는 마일스톤의 경우 편집창이 뜨지 않던 문제 해결

- String 타입에서 날짜값을 읽어오지 못해서 발생한 문제. 서버에서 반환되는 값은 String이므로, Date타입으로 변경해서 바인딩하도록 수정함

seedspirit committed:
Merge branch 'dev' into feat/#45

jellyyelly reviewed (approved):
고생하셨습니다!

seedspirit committed:
[feat/#45] feat: 충돌 해결

seedspirit committed:
Merge branch 'feat/#45' of https://github.com/codesquad-masters2024-team09/issue-tracker into feat/#45
